### PR TITLE
Convert log file prints to debug prints

### LINKED ISSUE #1539: [FEATURE REQUEST] Reduce lines printed by default to log file
The GEOS-Chem log file includes both GEOS-Chem and HEMCO prints starting in 14.0. This has made the already large file even larger. It would be good to go through what is printed and determine what is actually needed by default. Everything else can be removed, either by including in the `debug_printout` option in `geoschem_config.yml` or in a higher value of `VERBOSE` or `WARNINGS` in HEMCO. We could also simplify the print options in HEMCO (e.g. consolidate the hierarchical `VERBOSE` and `WARNINGS` into a single on/off switch) but I will make a separate feature request for that in HEMCO.

### PR DESCRIPTION
**Pull request by yantosca**
This is the corresponding PR to #1539.  We are converting print statements to debug print statements (toggled only when `verbose: true` is specified in `geoschem_config.yml`.

Closes #1539

### CODE REVIEW

**Comment by yantosca** [comment]
Integration tests pending.  

Also note: Corresponding updates have been made in https://github.com/geoschem/HEMCO/pull/182.  Namely, `Verbose: true` is now used to toggle all Verbose and Warnings output.

**Comment by yantosca** [comment]
All GEOS-Chem Classic integration tests passed
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

Using 24 OpenMP threads
Number of execution tests: 48
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_025x03125_CH4_geosfp_47L_na...................Execute Simulation.....PASS
gc_025x03125_fullchem_geosfp_47L_na..............Execute Simulation.....PASS
gc_05x0625_CH4_merra2_47L_na.....................Execute Simulation.....PASS
gc_05x0625_fullchem_merra2_47L_na................Execute Simulation.....PASS
gc_2x25_aerosol_merra2...........................Execute Simulation.....PASS
gc_2x25_CH4_merra2...............................Execute Simulation.....PASS
gc_2x25_CO2_merra2...............................Execute Simulation.....PASS
gc_2x25_fullchem_merra2..........................Execute Simulation.....PASS
gc_2x25_Hg_merra2................................Execute Simulation.....PASS
gc_2x25_metals_merra2............................Execute Simulation.....PASS
gc_2x25_POPs_BaP_merra2..........................Execute Simulation.....PASS
gc_2x25_tagCH4_merra2............................Execute Simulation.....PASS
gc_2x25_tagCO_merra2.............................Execute Simulation.....PASS
gc_2x25_tagO3_merra2.............................Execute Simulation.....PASS
gc_2x25_TransportTracers_merra2..................Execute Simulation.....PASS
gc_2x25_TransportTracers_merra2_LuoWd............Execute Simulation.....PASS
gc_4x5_aerosol_geosfp............................Execute Simulation.....PASS
gc_4x5_aerosol_merra2............................Execute Simulation.....PASS
gc_4x5_CH4_geosfp................................Execute Simulation.....PASS
gc_4x5_CH4_merra2................................Execute Simulation.....PASS
gc_4x5_fullchem_aciduptake_merra2................Execute Simulation.....PASS
gc_4x5_fullchem_APM_merra2.......................Execute Simulation.....PASS
gc_4x5_fullchem_benchmark_merra2.................Execute Simulation.....PASS
gc_4x5_fullchem_complexSOA_merra2................Execute Simulation.....PASS
gc_4x5_fullchem_complexSOA_SVPOA_merra2..........Execute Simulation.....PASS
gc_4x5_fullchem_geosfp...........................Execute Simulation.....PASS
gc_4x5_fullchem_marinePOA_merra2.................Execute Simulation.....PASS
gc_4x5_fullchem_merra2...........................Execute Simulation.....PASS
gc_4x5_fullchem_merra2_47L.......................Execute Simulation.....PASS
gc_4x5_fullchem_merra2_LuoWd.....................Execute Simulation.....PASS
gc_4x5_fullchem_RRTMG_merra2.....................Execute Simulation.....PASS
gc_4x5_fullchem_TOMAS15_merra2_47L...............Execute Simulation.....PASS
gc_4x5_fullchem_TOMAS40_merra2_47L...............Execute Simulation.....PASS
gc_4x5_Hg_geosfp.................................Execute Simulation.....PASS
gc_4x5_Hg_merra2.................................Execute Simulation.....PASS
gc_4x5_metals_merra2.............................Execute Simulation.....PASS
gc_4x5_POPs_BaP_geosfp...........................Execute Simulation.....PASS
gc_4x5_POPs_BaP_merra2...........................Execute Simulation.....PASS
gc_4x5_tagCH4_geosfp.............................Execute Simulation.....PASS
gc_4x5_tagCH4_merra2.............................Execute Simulation.....PASS
gc_4x5_tagCO_geosfp..............................Execute Simulation.....PASS
gc_4x5_tagCO_merra2..............................Execute Simulation.....PASS
gc_4x5_tagO3_geosfp..............................Execute Simulation.....PASS
gc_4x5_tagO3_merra2..............................Execute Simulation.....PASS
gc_4x5_TransportTracers_geosfp...................Execute Simulation.....PASS
gc_4x5_TransportTracers_geosfp_LuoWd.............Execute Simulation.....PASS
gc_4x5_TransportTracers_merra2...................Execute Simulation.....PASS
gc_4x5_TransportTracers_merra2_LuoWd.............Execute Simulation.....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 48
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by yantosca** [comment]
Samples of log files with and without debug printout activated:
- With debug printout: [GC.log.verbose.txt](https://github.com/geoschem/geos-chem/files/10211402/GC.log.verbose.txt)

- Without debug printout: [HEMCO_sa.log.noverbose.txt](https://github.com/geoschem/geos-chem/files/10211401/HEMCO_sa.log.noverbose.txt)

**Comment by yantosca** [comment]
Also tagging @jimmielin.  Long story short, we changed "Verbose" and "Warnings" to a single "Verbose: true" flag to toggle debug output.  I didn't touch the CESM config files but maybe I should.  When you have time, could let us know if you see anything that would break CESM-GC functionality?

**Review by msulprizio — COMMENTED** [review]
In some places we use:

```
IF ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
``` 

and in others we define and use prtDebug:

```
prtDebug = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
...
IF ( prtDebug ) THEN
```

Could we use prtDebug consistently throughout GEOS-Chem?

File: History/history_mod.F90 [hunk 1]
```diff
@@ -1796,7 +1795,7 @@ SUBROUTINE History_ReadCollectionData( Input_Opt,  State_Chm, State_Diag,  &
        DO C = 1, CollectionCount
           print*, 'Collection        ', TRIM( CollectionName       (C) )
           print*, '  -> FileName     ', TRIM( CollectionFileName   (C) )
-          print*, '  -> Format       ', TRIM( CollectionFormat     (C) )
+         !print*, '  -> Format       ', TRIM( CollectionFormat     (C) )
```
Comment (msulprizio) [inline History/history_mod.F90:1798]: Remove line rather than commenting out?

**Review by msulprizio — CHANGES_REQUESTED** [review]

**Comment by yantosca** [comment]
> In some places we use:
> 
> ```
> IF ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
> ```
> 
> and in others we define and use prtDebug:
> 
> ```
> prtDebug = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
> ...
> IF ( prtDebug ) THEN
> ```
> 
> Could we use prtDebug consistently throughout GEOS-Chem?

Yes, we could even use `verbose` instead of `prtDebug`.

If it is possible, I will investigate if it is possible to add a field such as `Input_Opt%Verbose` which could be set as:

```F90
Input_Opt%Verbose = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
```
I believe the "amIRoot" test is done in the initialization section so this approach would save on a lot of logical comparisons.

We can also change the `Input_Opt%LPRT` to `Input_Opt%doVerboseOutput` or something along those lines.  To make a distinction between "here is the switch that toggles verbose output" vs. "here is the switch that contains the result of whether or not we have to call this print statement".  I'll work on this.

@jimmielin: Do you foresee any issues with such an approach for either CESM-GC or WRF-GC?

**Comment by yantosca** [comment]
> Just wanted to confirm - what is the behavior of HEMCO if an old configuration file is used? I can update HEMCO-CESM with the new code easily but I want to make sure the model does not crash when:
> 
> (1) old `Verbose: 2` is used. A (2) a now unknown option `Warnings: 1` is included in the configuration file

I haven't tried this yet, but it should be straightforward to wriite logic for this.  We could force a "Verbose > 0" or "Warnings > 0" to "Verbose: true" if you think that's the best way forward (with maybe a print statement saying that debug output is now on).  

Do you think a lot of CESM-GC users will compilain if they get the error?  If so we can just force to `verbose: true` and call it a day.

**Comment by jimmielin** [comment]
> @jimmielin: Do you foresee any issues with such an approach for either CESM-GC or WRF-GC?

WRF-GC is able to separate the output into different CPUs. Our users sometime report that errors only appear in the other CPUs, so I would suggest having it like

```fortran
#if defined( MODEL_WRF )
    Input_Opt%Verbose = Input_Opt%LPRT
#else
    Input_Opt%Verbose = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
#endif
```

> I haven't tried this yet, but it should be straightforward to wriite logic for this. We could force a Verbose > 0 or Warnings > 0 to Verbose: true if you think that's the best way forward (with maybe a print statement saying that debug output is now on).

Do you think a lot of CESM-GC users will compilain if they get the error? If so we can just force to verbose: true and call it a day.

Let's just force it to `Verbose: true` if `Verbose > 0 or Warnings > 0`. We can make a printout saying that "Numbered verbose and warnings options are deprecated. Please use Verbose: true or false for controlling verbose output". Sorry if it will give a bit of trouble parsing the configuration file, but it will be nice to have this for backwards compatibility. They shouldn't mind the warning as long as the code has well-defined behavior.

Thanks Bob!

**Comment by yantosca** [comment]
Thanks @jimmielin, am working on these updates now.

**Comment by lizziel** [comment]
@jimmielin, the `HEMCO_Config.rc` file is going to be tied to the GEOS-Chem version used in CESM. So when GEOS-Chem is updated in CESM with this update then `run/CESM/HEMCO_Config.rc` will automatically be put into people's `Buildconf/camconf` directory. Users can swap it out for an older `HEMCO_Config.rc` if they want, but we should recommend that they always compare the two and make adjustments as needed since we do regularly change things in that file, e.g. extensions menus. As a practice we don't add code to GEOS-Chem to warn that config file options are deprecated since they change so frequency. So all this to say I'd vote against having special handling for if users are using an older config file.

**Comment by jimmielin** [comment]
Thanks @lizziel. I am aware of this handling in GEOS-Chem within CESM. However CAM-chem users maintain separate configuration files that are not updated together with HEMCO or the interface. The expectation is that it would be updated together with CAM-chem compsets in the future.

A breaking change to the configuration file that is not as part of a major version release may set a bad precedent, especially if it is not related to emission entries and is expected to be rarely updated.

**Comment by yantosca** [comment]
@lizziel @jimmielin: Maybe we could leave in the "deprecated" warnings as a silent (undocumented) feature until the compset gets updated.  It's probably easier to change GEOS-Chem than CESM or CAM.

**Comment by yantosca** [comment]
> In some places we use:
> 
> ```
> IF ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
> ```
> 
> and in others we define and use prtDebug:
> 
> ```
> prtDebug = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
> ...
> IF ( prtDebug ) THEN
> ```
> 
> Could we use prtDebug consistently throughout GEOS-Chem?

We now use `IF ( Input_Opt%VerboseAndRoot )` to toggle verbose output.  See commits 5f9a83e72663c94e34602e82196982db7a61ade7, 5f9a83e72663c94e34602e82196982db7a61ade7, 5f9a83e72663c94e34602e82196982db7a61ade7, and e713783594df2295433861effa03b65c74c4af49

**Comment by yantosca** [comment]
@lizziel: Here are the output from 1-hr GCHP runs with verbose output on and off:

- [gchp.20190701_0000z.log.noverbose.txt](https://github.com/geoschem/geos-chem/files/10239039/gchp.20190701_0000z.log.noverbose.txt) (148 kB)

- [gchp.20190701_0100z.log.verbose.txt](https://github.com/geoschem/geos-chem/files/10239046/gchp.20190701_0100z.log.verbose.txt) (3869 kb)

File: Headers/state_chm_mod.F90 [hunk 2]
```diff
@@ -2322,11 +2320,11 @@ SUBROUTINE Init_Mapping_Vectors( Input_Opt, State_Chm, RC )
...
-          ! Print to screen
-          IF ( Input_Opt%amIRoot ) THEN
-             WRITE( 6, 100 ) ThisSpc%ModelId, ThisSpc%Name
- 100         FORMAT( I3, 2x, A31 )
-          ENDIF
+!          ! Print to screen
+!          IF ( Input_Opt%amIRoot ) THEN
+!             WRITE( 6, 100 ) ThisSpc%ModelId, ThisSpc%Name
+! 100         FORMAT( I3, 2x, A31 )
+!          ENDIF
```
[the remark this answers is not in the document]
Comment (yantosca) [reply Headers/state_chm_mod.F90]: Now removed in commit b349dde6208df7ce892cbd9f012ba5327a1083a5

File: History/history_mod.F90 (hunk 1 above)
Comment (yantosca) [reply History/history_mod.F90]: Now removed in commit b349dde6208df7ce892cbd9f012ba5327a1083a5.  Actually, we no longer use the CollectionFormat metadata, so I've removed that array from `history_mod.F90`.

**Comment by yantosca** [comment]
> WRF-GC is able to separate the output into different CPUs. Our users sometime report that errors only appear in the other CPUs, so I would suggest having it like
> 
> ```fortran
> #if defined( MODEL_WRF )
>     Input_Opt%Verbose = Input_Opt%LPRT
> #else
>     Input_Opt%Verbose = ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
> #endif
> ```

This is now done in commit e713783594df2295433861effa03b65c74c4af49

**Comment by lizziel** [comment]
Thanks for posting the log files. It's hard to see what the GEOS-Chem debug prints are since the HEMCO verbose is so dominant in the log file. I see this as printed 6 times, but haven't seen any other, so maybe it's okay.
```
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
 STUBBED in CPL ReadRestart
 DEBUG: no cpl restart found, producing partial month
```
Regarding @jimmielin's point about errors on non-root threads, I wonder if we should have a second toggle for verbose on root only. If verbose is on and verbose on root only is false then the prints would be for all CPUs. Would that be simple to implement?

**Comment by lizziel** [comment]
To answer my own question, I think it a verbose on root only could simply be used to modify input_opt%verbose, like this:
if ( Input_Opt%VerboseRootOnly ) then
    Input_Opt%Verbose =  ( Input_Opt%amIRoot .and. Input_Opt%LPRT )
else
    Input_Opt%Verbose = Input_Opt%LPRT 
endif

**Comment by yantosca** [comment]
@lizziel, @jimmielin @msulprizio:  Maybe we can make this more general and do:

```yaml
verbose:
  activate: true 
  on_cores: root | all
```

**Comment by yantosca** [comment]
@msulprizio @lizziel @jimmielin, here's what I've done (example from GCClassic):
```yaml
simulation:
  name: fullchem
  start_date: [20190701, 000000]
  end_date: [20190701, 002000]
  root_data_dir: /n/holyscratch01/external_repos/GEOS-CHEM/gcgrid/data/ExtData
  met_field: MERRA2
  species_database_file: ./species_database.yml
  verbose:
    activate: false
    on_cores: root               # Accepted values: root, all
  use_gcclassic_timers: false
```
And then in input_mod.F90:
https://github.com/geoschem/geos-chem/blob/1712ad4ad1dc3cad4d2c5358f329781865f4c3fa/GeosCore/input_mod.F90#L537-L577

which gives this output:
```console
===> SIMULATION START TIME: 2022/12/22 14:26 <===

===============================================================================
G E O S - C H E M   U S E R   I N P U T

READ_INPUT_FILE: Opening ./geoschem_config.yml

SIMULATION SETTINGS
-------------------
Simulation name             : fullchem
CHEM_INPUTS directory       : /n/holyscratch01/external_repos/GEOS-CHEM/gcgrid/data/ExtData/CHEM_INPUTS/
Species database file       : ./species_database.yml
Turn on verbose output      :     F
Verbose output printed on   : root core only
Start time of run           : 20190701 000000
End time of run             : 20190701 002000
Data Directory              : /n/holyscratch01/external_repos/GEOS-CHEM/gcgrid/data/ExtData/
Meteorology field           : MERRA2
Turn on GEOS-Chem timers    :     F
```
I will add a tag in HEMCO_Config.rc to toggle output on root core or all cores.

**Comment by yantosca** [comment]
I have merged the logfile cleanup updates to GEOS-Chem 14.1.0 and HEMCO 3.6.0, and have fixed post-merge issues.  Integration tests are now running.

**Pull request merged**